scio-test/src/main/scala/com/spotify/scio/testing/ParquetTestUtils.scala

```diff
@@ -0,0 +1,174 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.scio.testing
+
+import magnolify.parquet._
+import org.apache.avro.Schema
+import org.apache.avro.generic.GenericRecord
+import org.apache.hadoop.conf.Configuration
+import org.apache.parquet.avro.{AvroParquetReader, AvroParquetWriter, AvroReadSupport}
+import org.apache.parquet.filter2.predicate.FilterPredicate
+import org.apache.parquet.hadoop.{ParquetInputFormat, ParquetReader, ParquetWriter}
+import org.apache.parquet.io._
+
+import java.io.{ByteArrayInputStream, ByteArrayOutputStream}
+import scala.reflect.ClassTag
+
+trait ParquetTestUtils {
+  sealed trait ImplementsFilter[T] {
+    def parquetFilter(filter: FilterPredicate): Iterable[T]
+  }
+
+  sealed trait ImplementsProject[T] {
+    def parquetProject(schema: Schema): Iterable[T]
+  }
+
+  case class ParquetMagnolifyHelpers[T: ParquetType: ClassTag](records: Iterable[T])
+      extends ImplementsFilter[T] {
+    override def parquetFilter(filter: FilterPredicate): Iterable[T] = {
+      val pt = implicitly[ParquetType[T]]
+
+      val configuration = new Configuration()
+      ParquetInputFormat.setFilterPredicate(configuration, filter)
+
+      roundtrip(
+        outputFile => pt.writeBuilder(outputFile).build(),
+        inputFile => pt.readBuilder(inputFile).withConf(configuration).build()
+      )(records)
+    }
+  }
+
+  case class ParquetAvroHelpers[T <: GenericRecord: ClassTag](records: Iterable[T])
+      extends ImplementsFilter[T]
+      with ImplementsProject[T] {
+    override def parquetProject(projection: Schema): Iterable[T] = {
+      val configuration = new Configuration()
+      AvroReadSupport.setRequestedProjection(configuration, projection)
+
+      roundtripAvro(records, configuration)
+    }
+
+    override def parquetFilter(filter: FilterPredicate): Iterable[T] = {
+      val configuration = new Configuration()
+      ParquetInputFormat.setFilterPredicate(configuration, filter)
+
+      roundtripAvro(records, configuration)
+    }
+
+    private def roundtripAvro(
+      records: Iterable[T],
+      readConfiguration: Configuration
+    ): Iterable[T] = {
+      records.headOption match {
+        case None =>
+          records // empty iterable
+        case Some(head) =>
+          val schema = head.getSchema
+
+          roundtrip(
+            outputFile => AvroParquetWriter.builder[T](outputFile).withSchema(schema).build(),
+            inputFile => AvroParquetReader.builder[T](inputFile).withConf(readConfiguration).build()
+          )(records)
+      }
+    }
+  }
+
+  // @Todo tensorflow helpers
```

Comment:
I think `scio-test` role is to provide testing framework for generic pipeline and should not containt specific IO/extension code.
I thinl `scio-parquet` is more adapted

Reply:
I'm wondering if this should even be in main so we apply those on test input.

Reply:
it could, that was my [initial implementation](https://github.com/spotify/scio/pull/5309/commits/1b4706e9007eb9efdc35447ac0ce58f2aa0aa1d4). Downside is that it changes the expected TestIO behavior, as @kellen pointed out, since `Parquet*IO`s would be transforming the provided input.

But maybe it's the "least bad" solution of these three possible options:

1) Refactor `scio-parquet` test suites so that `scio-test` can depend on `scio-parquet` in order to access the Tensorflow Parquet Writer in the new test util
2) Create a new artifact `scio-parquet-test` with the new test util
3) Move the test util logic to `Parquet*IO`'s (ParquetAvroIO, ParquetExampleIO, ParquetTypeIO) `readTest` so that it's applied to all test Parquet input